### PR TITLE
Compare Nullable modifiers while comparing type symbols by default.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

**Comment by AlekseyTs** [comment]
Test windows_release_vs-integration_prtest please

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs [hunk 1]
```diff
@@ -2905,7 +2905,7 @@ private static void GetAllCandidates(Dictionary<TypeSymbolWithAnnotations, TypeS
                     {
                         candidates.Remove(candidate);
                     }
-                    else if (bound.Equals(candidate, TypeCompareKind.IgnoreDynamicAndTupleNames))
+                    else if (bound.Equals(candidate, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes))
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs]: This seems to be used relatively often. Might be worth adding an explicit option in the enum.

File: src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs [hunk 2]
```diff
@@ -512,7 +512,7 @@ public int GetHashCode(Symbol member)
                 }
 
                 if (_considerReturnType && member.GetMemberArity() == 0 && 
-                    (_typeComparison & TypeCompareKind.IgnoreCustomModifiersAndArraySizesAndLowerBounds) == 0) // If it is generic, then type argument might be in return type.
+                    (_typeComparison & TypeCompareKind.AllIgnoreOptions) == 0) // If it is generic, then type argument might be in return type.
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: Isn't this ignoring new things besides nullability? Why the change?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs]: > Isn't this ignoring new things besides nullability? Why the change?

At this point I believe the change is necessary to ensure correctness because at the moment some implementations of GetHashCode on symbols actually sensitive to things that we are supposed to ignore. See [https://github.com/dotnet/roslyn/issues/30673](https://github.com/dotnet/roslyn/issues/30673). I'll add a note to that issue that once we ensure that all GetHashCode implementations do the right thing, this condition probably can be removed.


---
In reply to: [228374243](https://github.com/dotnet/roslyn/pull/30770#discussion_r228374243) [](ancestors = 228374243)

**Comment by AlekseyTs** [comment]
@cston Please review

**Comment by cston** [comment]
>         ConsiderEverything = 0, // Perhaps rename since this does not consider nullable modifiers.

Can be removed.

---
Refers to: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs:13 in d74070e. [](commit_id = d74070eeeb39bf8b4502f22c421f4a9470986f86, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 3]
```diff
@@ -50800,5 +50800,323 @@ public class EqualityComparer<T>
...
+                // (30,22): error CS0540: 'Outer<T>.Inner<U>.Derived4.Derived6.Outer<T>.Inner<U>.Interface<U, T>.Method<K>(T, U[], List<U>, Dictionary<T, K>)': containing type does not implement interface 'Outer<T>.Inner<U>.Interface<U, T>'
+                //                 void Inner<U>.Interface<U, T>.Method<K>(T a, U[] b, List<U> c, Dictionary<T, K> D)
+                Diagnostic(ErrorCode.ERR_ClassDoesntImplementInterface, "Inner<U>.Interface<U, T>").WithArguments("Outer<T>.Inner<U>.Derived4.Derived6.Outer<T>.Inner<U>.Interface<U, T>.Method<K>(T, U[], System.Collections.Generic.List<U>, System.Collections.Generic.Dictionary<T, K>)", "Outer<T>.Inner<U>.Interface<U, T>").WithLocation(30, 22)
+                );
+        }
+
+        [Fact]
+        [WorkItem(30677, "https://github.com/dotnet/roslyn/issues/30677")]
+        public void TestErrorsImplementingGenericNestedInterfaces_Explicit_IncorrectPartialQualification()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >Incorrect [](start = 75, length = 9)

What does `Incorrect` refer to?

**Comment by cston** [comment]
Consider targeting master rather than features/NullableReferenceTypes.

**Comment by AlekseyTs** [comment]
>         ConsiderEverything = 0, // Perhaps rename since this does not consider nullable modifiers.

> Can be removed.

I think it is used.

---
In reply to: [433565246](https://github.com/dotnet/roslyn/pull/30770#issuecomment-433565246) [](ancestors = 433565246)

---
Refers to: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs:13 in d74070e. [](commit_id = d74070eeeb39bf8b4502f22c421f4a9470986f86, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: > What does Incorrect refer to?

This unit-test is a clone (possibly partial) of the other unit-test, the name is inherited from the original unit-test.

---
In reply to: [228689077](https://github.com/dotnet/roslyn/pull/30770#discussion_r228689077) [](ancestors = 228689077)

**Comment by cston** [comment]
>         ConsiderEverything = 0, // Perhaps rename since this does not consider nullable modifiers.

Sorry, I was referring to the comment. I think the comment can be removed.

---
In reply to: [433985024](https://github.com/dotnet/roslyn/pull/30770#issuecomment-433985024) [](ancestors = 433985024,433565246)

---
Refers to: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs:13 in d74070e. [](commit_id = d74070eeeb39bf8b4502f22c421f4a9470986f86, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         ConsiderEverything = 0, // Perhaps rename since this does not consider nullable modifiers.

> I was referring to the comment. I think the comment can be removed.

Sure. Thanks for the clarification. Will remove

---
In reply to: [433987578](https://github.com/dotnet/roslyn/pull/30770#issuecomment-433987578) [](ancestors = 433987578,433985024,433565246)

---
Refers to: src/Compilers/Core/Portable/Symbols/TypeCompareKind.cs:13 in d74070e. [](commit_id = d74070eeeb39bf8b4502f22c421f4a9470986f86, deletion_comment = False)
